### PR TITLE
feat(deployment): add basic Deployment resource

### PR DESCRIPTION
**Pull request by skyscrapr**
## Summary

Adds Deployment resource with basic fields. More complete support for Deployments will come in future PRs (see https://github.com/PrefectHQ/terraform-provider-prefect/issues/238).

Related to https://github.com/PrefectHQ/terraform-provider-prefect/issues/238

Depends on https://github.com/PrefectHQ/terraform-provider-prefect/pull/201.

## Notes

- API docs: https://app.prefect.cloud/api/docs#tag/Deployments
- This PR is already big, so we're implementing some core fields here and [the rest](https://github.com/PrefectHQ/terraform-provider-prefect/issues/238) will be implemented in future PR(s)
- CI `Acceptance tests` are sometimes failing with `503 service unavailable` (and sometimes [passing](https://github.com/PrefectHQ/terraform-provider-prefect/actions/runs/10379690891/job/28738722369)), but I tested the newly-added test by hand and it passed:

   ```
   $ go test ./... -count=1 -v -run TestAccResource_deployment
   ?       github.com/prefecthq/terraform-provider-prefect [no test files]
   .. etc ...
   ?       github.com/prefecthq/terraform-provider-prefect/internal/utils  [no test files]
   testing: warning: no tests to run
   PASS
   ok      github.com/prefecthq/terraform-provider-prefect/internal/provider/datasources   0.339s [no tests to run]
   === RUN   TestAccResource_deployment
   === PAUSE TestAccResource_deployment
   === CONT  TestAccResource_deployment
   --- PASS: TestAccResource_deployment (3.66s)
   PASS
   ok      github.com/prefecthq/terraform-provider-prefect/internal/provider/resources     4.252s
   ```

## To do

- [x] Run a basic test to create and destroy a Flow + Deployment
- [x] Implement necessary attributes (others can come in a separate PR)
- [x] Ensure current acceptance tests pass
- [x] Expand acceptance tests as necessary

### CODE REVIEW

**Comment by skyscrapr** [comment]
Deployments are dependent on flows
You cannot create a deployment without a flow_id. 

I split flows into #201. But I'm unsure. How to create this PR if it requires another PR to be merged first. The branch will not pass tests. 

If you explain how to do this, happy to update the PR and remove flows.

I have a third PR for deployment access that I haven't opened since it is depending on deployments and only makes the problem worse.

**Comment by skyscrapr** [comment]
The other option is to merge #201 and that should remove flows from #202 automatically.

**Comment by mitchnielsen** [comment]
Yeah, I was thinking we could set the Deployment PR's branch target to this Flow PR, but I don't think there's a way to do that with forks. 

Let's mark the Deployment PR as draft for now, and when this PR merges we'll revisit that one. That way, any changes we make here don't need to also be applied to your Deployment PR.

**Comment by mitchnielsen** [comment]
Not sure why the [Terraform docs job](https://github.com/PrefectHQ/terraform-provider-prefect/actions/runs/9637652715/job/26957379060?pr=202) is failing.

**Comment by tehioant** [comment]
@skyscrapr do you need help finishing this pr ? I would love to see this feature.
If I can help in any way 🙂

**Comment by mitchnielsen** [comment]
Hi @skyscrapr, thanks again for submitting this. I'm catching up from PTO from a move 📦 

I'd like to take a look at this one this week. Could you please address the conflicts?

**Comment by mitchnielsen** [comment]
@tehioant - thank you for reaching out and offering your help 🤝 Once @skyscrapr resolves the conflicts I'll dive in and test out the changes. If we need any extra hands I'll reach out!

**Comment by parkedwards** [comment]
hey @skyscrapr - if you're open to it, we'd be be happy to help help take this PR home + squash any of the remaining merge conflicts.  we can either push to your forked repo on this branch, or re-create a branch with these changes.  let us know

**Comment by skyscrapr** [comment]
@parkedwards  yes please go ahead. This PR contains basic functionality and still needs to be extended to support schedules etc. However it would be good to get this merged so that we can work on adding support for schedules etc. I'm currently tied up in my day job. :-) But can help support later this week.

**Comment by skyscrapr** [comment]
Best to push to my forked branch please as this will allow us to continue working on deployment access and schedules.

**Comment by parkedwards** [comment]
@skyscrapr totally makes sense. we'll push to this branch. thank you for all your work on this so far - between me, you and @mitchnielsen we'll be able to steer this home

**Comment by skyscrapr** [comment]
This initial PR adds support for deployments but does not support all attributes. 

However if you can merge this change first then deployment access resource PR can be opened while support for more attributes can be added in parallel. 

Let me know if that makes sense.

**Comment by mitchnielsen** [comment]
> This initial PR adds support for deployments but does not support all attributes.
> 
> However if you can merge this change first then deployment access resource PR can be opened while support for more attributes can be added in parallel.
> 
> Let me know if that makes sense.

That seems fair to me. There are [quite a few attributes](https://github.com/PrefectHQ/nebula/blob/3836c49368641ffb97876d980cd90ada0168898c/src/prefect/server/schemas/core.py#L453) so I'm open to merging it with the essentials and we can add more in a follow-up PR.

**Comment by mitchnielsen** [comment]
Getting close. Working on tests, and trying to figure out why updating any fields on the Deployment gives a `403`.

File: docs/resources/flow.md [hunk 1]
```diff
@@ -50,5 +50,5 @@ Import is supported using the following syntax:
 
 ```shell
 # Prefect Flows can be imported via flow_id,workspace_id
-terraform import prefect_variable.example 00000000-0000-0000-0000-000000000000,00000000-0000-0000-0000-000000000000
+terraform import prefect_flow.example 00000000-0000-0000-0000-000000000000,00000000-0000-0000-0000-000000000000
```
Comment (mitchnielsen) [inline docs/resources/flow.md:53]: This is just a drive-by fix I saw when working on testing Deployments with Flows.

File: internal/api/deployments.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	WorkPoolName           string   `json:"work_pool_name,omitempty"`
+	WorkQueueName          string   `json:"work_queue_name,omitempty"`
+}
+
+// DeploymentFilter defines the search filter payload
+// when searching for deployements by name.
+// example request payload:
+// {"deployments": {"handle": {"any_": ["test"]}}}.
+type DeploymentFilter struct {
+}
```
Comment (mitchnielsen) [inline internal/api/deployments.go:73]: There are some extra structs here and below that aren't implemented yet, but I didn't want to delete them and have to recreate them later. Happy to do that though if desired.

File: internal/provider/helpers/template.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	"text/template"
+)
+
+// RenderTemplate is a helper function to take a template and some configuration and
+// return the result as a string.
+//
+// Where:
+//   - tpl is a string like "Hello, {{.Name}}"
+//   - data is a struct like `mydata{Name: "World"}`
+func RenderTemplate(tpl string, data interface{}) string {
```
Comment (mitchnielsen) [inline internal/provider/helpers/template.go:14]: The Terraform manifest building got really unwieldy with all of the positional arguments to `Sprintf`, so I used Go's `text/template` here so it's a bit easier to read and maintain.

File: docs/resources/flow.md (hunk 1 above)
Comment (parkedwards) [reply docs/resources/flow.md:53]: copy pasta strikes again

File: examples/resources/prefect_deployment/resource.tf [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+resource "prefect_flow" "flow" {
+  name         = "my-flow"
+  workspace_id = prefect_workspace.workspace.id
+  tags         = ["tf-test"]
+}
+
+resource "prefect_deployment" "deployment" {
+  name                     = "%s"
```
Comment (parkedwards) [inline examples/resources/prefect_deployment/resource.tf:13]:
```suggestion
  name                     = "my-deployment"
```

File: internal/api/deployments.go (hunk 2 above)
Comment (parkedwards) [reply internal/api/deployments.go:73]: happy to keep them in

File: internal/client/deployments.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+	var deployments []*api.Deployment
+	if err := json.NewDecoder(resp.Body).Decode(&deployments); err != nil {
+		return nil, fmt.Errorf("failed to decode response: %w", err)
+	}
+
+	return deployments, nil
+}
+
+// Get returns details for a Workspace by ID.
```
Comment (parkedwards) [inline internal/client/deployments.go:107]:
```suggestion
// Get returns details for a Deployment by ID.
```

**Review by parkedwards — APPROVED** [review]

**Comment by mitchnielsen** [comment]
Thanks for getting this one off the ground @skyscrapr 🤝
